Add gravity-compensated kinematic equations and feedback control (#419)
gamemanual0/gm0: source/docs/software/concepts/kinematics.rst

   v_{br} = v_f - v_s + (2r_b \cdot \omega)

   v_{fr} = v_f + v_s + (2r_b \cdot \omega)

Manipulators
--------------

.. _gravity-compensation:

Gravity Compensation
^^^^^^^^^^^^^^^^^^^^

Often in FTC, teams have arms that swing out around an axis. Controlling these arms requires a bit of thought as depending on the angle they're at, the effect of gravity drastically changes.

Our first step is defining a reference frame. Our recommendation is to define zero as straight up and down, as this is when gravity doesn't have an effect on your arm. Other reference frames will still work, but the trigonometry won't be as straightforward.

.. figure:: ./images/kinematics/reference_frame.svg
   :alt: A diagram of a robot with a long arm, with 0 degrees marked up and down

>    :align: center
>    :width: 400px

Reviewer: ```suggestion
```

Author: I don't know how this behaves on your computer, but on mine (firefox) implicitly sizing the image makes it super small. Thankfully, 400px is the intended width, and will appropriately resize to be smaller upon screen changes.

Reviewer: I learned something about the HTML today: 
![image](https://github.com/gamemanual0/gm0/assets/16172089/61312abf-2fe3-4a64-a6b3-11f6b9d09d35)

We use a custom sphinx extension to set default image width/alignment, and it sets the sphinx image width attribute to 25 em, which works for PNGs/JPEGs as it gets set as the CSS width property, but for some reason on SVGs it gets set as the width tag on the `<img>` element, and so becomes 25 pixels wide. I have since fixed this, by making it set the image width attribute to 400 (as in, 400 pixels).

>    Image courtesy of 11329 ICE Robotics

Reviewer: ```suggestion
   11329 I.C.E. Robotics
```

In a reference frame like this, the relative force of gravity will be equal to the sin of the angle. This makes sense as when the sin function is evaluated at zero degrees, it returns 0, while at 90 degrees where the effect of gravity is at its greatest, it returns 1.

.. math::

>    F_g = g * \sin{\theta}

Reviewer: ```suggestion
   F_g = g\sin{\theta}
```

> Assuming you have an encoder on your arm, you can find :math:`\theta` pretty easily using the following code.

Reviewer: ```suggestion
Assuming you have an encoder on your arm, you can find :math:`\theta` (the angle of the arm relative to the vertical) using something similar to the following pseudocode.
```

> .. note:: DEGREE_PER_TICK can be found by taking 360 divided by your encoder resolution all multiplied by your gear ratio. A GoBuilda 5202 motor with a gear ratio of 1:1 will have a TICK_PER_REV of 751.8, and a DEGREE_PER_TICK of 0.47885075818.

Reviewer: ```suggestion
.. note:: DEGREE_PER_TICK can be found by taking 360 divided by your encoder resolution all multiplied by your gear ratio. For example, for a `19.2:1 goBILDA Yellow Jacket <https://www.gobilda.com/5203-series-yellow-jacket-planetary-gear-motor-19-2-1-ratio-24mm-length-8mm-rex-shaft-312-rpm-3-3-5v-encoder/>`_, there are 537.7 ticks per revolution, and so :math:`360 / 537.7` degrees per tick.
```

> .. note:: DEGREE_PER_TICK can be found by taking 360 divided by your encoder resolution all multiplied by your gear ratio. A GoBuilda 5202 motor with a gear ratio of 1:1 will have a TICK_PER_REV of 751.8, and a DEGREE_PER_TICK of 0.47885075818.

Reviewer: ```suggestion
.. note:: DEGREE_PER_TICK can be found by taking 360 divided by your encoder resolution all multiplied by your gear ratio. For example, for a `19.2:1 goBILDA Yellow Jacket <https://www.gobilda.com/5203-series-yellow-jacket-planetary-gear-motor-19-2-1-ratio-24mm-length-8mm-rex-shaft-312-rpm-3-3-5v-encoder/>`_, there are 537.7 ticks per revolution, and so :math:`\frac{360}{537.7}` degrees per tick at the gearbox output shaft. If there was a 2:1 gear ratio after that, then the ticks per revolution would instead be :math:`\frac{360}{2 \times 537.7}`
```
